### PR TITLE
added java homework fruit shop solution

### PR DESCRIPTION
**Pull request by Basilius-Third**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/Reader.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface Reader {
+    List<String> readFromFile(String fileName);
```
Comment (Optipist) [inline src/main/java/core/basesyntax/service/Reader.java:6]: Maybe just change name to read?

File: src/main/java/core/basesyntax/service/impl/FileReaderImpl.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+public class FileReaderImpl implements Reader {
+    @Override
+    public List<String> readFromFile(String fileName) {
+        File fruitStatistics = new File(fileName);
+        List<String> lines = new ArrayList<>();
+        try (BufferedReader fruitReader = new BufferedReader(new FileReader(fruitStatistics))) {
+            String line = fruitReader.readLine();
+            while (line != null) {
+                lines.add(line);
+                line = fruitReader.readLine();
```
Comment (Optipist) [inline src/main/java/core/basesyntax/service/impl/FileReaderImpl.java:20]: You can use class Files and method readAllLines()

**Review by Optipist — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Main.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    public static void main(String[] args) {
+        Map<String, OperationHandler> handlers = new HashMap<>();
+        handlers.put(OPERATION_BALANCE, new BalanceOperationHandler());
+        handlers.put(OPERATION_SUPPLY, new AddOperationHandler());
+        handlers.put(OPERATION_PURCHASE, new PurchaseOperationHandler());
+        handlers.put(OPERATION_RETURN, new AddOperationHandler());
+
+        List<String> linesFromFile = new FileReaderImpl()
+                .read("src/main/resources/fruit_file.csv");
```
Comment (PashaTerletskyi) [inline src/main/java/core/basesyntax/Main.java:33]: "src/main/resources/fruit_file.csv" remove to constant

File: src/main/java/core/basesyntax/Main.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        linesFromFile.stream()
+                .skip(1)
+                .map(parser::parseLine)
+                .forEach(operation -> new OperationStrategy(handlers)
+                        .get(operation.getOperation())
+                        .apply(operation));
+
+        String report = new FruitServiceImpl().createReport();
+        new FileWriterImpl()
+                .write("src/main/resources/fruit_report_from_" + LocalDate.now(), report);
```
Comment (PashaTerletskyi) [inline src/main/java/core/basesyntax/Main.java:44]: "src/main/resources/fruit_report_from_" remove to constant

**Review by Tamara-Paskalova — APPROVED** [review]

File: src/main/java/core/basesyntax/strategy/PurchaseOperationHandler.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    @Override
+    public int apply(Transaction transactionDto) {
+        Fruit fruit = new Fruit(transactionDto.getName());
+        int currentQuantity = Storage.storage.getOrDefault(fruit, 0);
+        int subtractQuantity = transactionDto.getQuantity();
+        if (currentQuantity - subtractQuantity < 0) {
+            throw new RuntimeException("Could not purchase! We don't have enough fruit["
+                    + fruit.getName() + "] in our storage!\n"
+                    + "Quantity left in storage - " + currentQuantity
+                    + "\nQuantity needed to buy - " + subtractQuantity);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/strategy/PurchaseOperationHandler.java:17]: next time we can use `String.format()`

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]
sorry, misclick) good job!

**Pull request closed without merging**